File: erlang/ch1/ex105.erl
%%%-------------------------------------------------------------------
%%% @author Yohan
%%% @copyright (C) 2018, <COMPANY>
%%% @doc
%%%
%%% @end
%%% Created : 27. Jul 2018 1:16 PM
%%%-------------------------------------------------------------------
-module(ex105).
-author("Yohan").

%% API
-export([test/2, p/0]).

p() -> p().

%%% try this> test(0, p). Does not do infinite recursion.
test(X, Y) ->
  if X == 0 -> 0;
    true -> Y
  end.

File: elixir/ch1/ex105.ex
# import_file("ex105.ex")

defmodule Ex105 do
  @moduledoc false

  def p(), do: p()

  def test(x, y) do
    cond do
      x == 0 -> 0
      true -> y
    end
  end

end

# try running Ex105.test(0, Ex105.p)
# this does not return

